src/runner_manager.py

```diff
@@ -0,0 +1,360 @@
+                    "Downloaded GitHub runner binary contains files: %s", ", ".join(f.getnames())
+                )
+            except tarfile.TarError:
+                logger.error("Failed to decompress downloaded GitHub runner binary.")
+                raise
+
+    def get_github_info(self) -> list[RunnerInfo]:
+        """Get information on the runners from GitHub.
+
+        Returns:
+            List of information from GitHub on runners.
+        """
+        remote_runners = self._get_runner_github_info()
+        return [
+            RunnerInfo(r.name, r.status == "online", r.status == "offline")
+            for r in remote_runners.values()
+        ]
+
+    def reconcile(self, quantity: int, resources: VirtualMachineResources) -> int:
+        """Bring runners in line with target.
+
+        Args:
+            quantity: Number of intended runners.
+            resources: Configuration of the virtual machine resources.
+
+        Returns:
+            Difference between intended runners and actual runners.
+        """
+        runners = self._get_runners()
+
+        # Clean up offline runners
+        offline_runners = [r for r in runners if not r.online]
+        if offline_runners:
+            runner_names = ", ".join(r.name for r in offline_runners)
+            logger.info("Cleaning up offline runners: %s", runner_names)
+
+            for runner in offline_runners:
+                runner.remove()
+
+        # Add/Remove runners to match the target quantity
+        online_runners = [r for r in runners if r.exist and r.online]
+        delta = quantity - len(online_runners)
+        if delta > 0:
+            logger.info("Getting registration token for GitHub runners.")
+
+            token: RegisterToken = {"token": None}
+            if isinstance(self.path, GitHubRepo):
+                token = self._github.actions.create_registration_token_for_repo(
+                    owner=self.path.owner, repo=self.path.repo
+                )
+            elif isinstance(self.path, GitHubOrg):
+                token = self._github.actions.create_registration_token_for_org(org=self.path.org)
+
+            logger.info("Adding %i additional runner(s)", delta)
+            for _ in range(delta):
+                config = RunnerConfig(
+                    self.app_name, self.path, self.proxies, self._generate_runner_name()
+                )
+                runner = Runner(self._github, self._jinja, self._lxd, config)
+                runner.create(
+                    self.image,
+                    resources,
+                    self.runner_bin_path,
+                    token.token,
+                )
+        elif delta < 0:
+            idle_runners = [runner for runner in online_runners if not runner.busy]
+            offset = min(-delta, len(idle_runners))
+            if offset != 0:
+                logger.info("Removing %i runner(s)", offset)
+                remove_runners = idle_runners[:offset]
```

Comment:
You are right this logic can cause removal of a busy runner.

The way to resolve this will involve the GitHub runner binary having an atomic operation that checking if the runner is currently busy and remove/unregister from GitHub if not busy.

There is two way to remove/unregister a runner, one method is executing the runner binary, the other is calling a RESTful API on GitHub. From the [documentation](https://docs.github.com/en/actions/hosting-your-own-runners/removing-self-hosted-runners#removing-a-runner-from-a-repository) is it unclear how these two method interacts with the runner being busy. I will need to do more research on this.

Reply:
I think this is not a high priority, as this is only possible due to configuration change on the number of the runners to maintain, which should occur infrequently.

Reply:
I agree, perhaps adding a TODO comment here, and/or creating a ticket on Jira to manage this case

Reply:
I'm not sure pylint will let you have TODO comments...

Reply:
@jdkandersson what do you think we should do here ?

Reply:
Maybe just create an issue in the repo, that way we have an easy way of telling all the problems we should fix

Reply:
I created an issue.